src/librbd/ImageCtx.h

```diff
@@ -68,6 +68,7 @@ namespace librbd {
                                         // a format librados can understand
     std::map<librados::snap_t, SnapInfo> snap_info;
     std::map<std::string, librados::snap_t> snap_ids;
+    std::map<cls::rbd::SnapshotNamespace, librados::snap_t> snap_namespace_ids;
```

Comment:
I am probably missing something, but you would need to do this regardless since you can have duplicate names between namespaces. There really would never be a valid case where you need to search by name only since you would need the namespace disambiguation. For the legacy snapshots, you would just default construct an image namespace and the snapshot name. For group snapshots, you would require the full group namespace to find the correct snapshot.

I would actually recommend breaking this piece of refactoring into its own commit within the PR since it is an atomic unit of work that needs to be done and would facilitate the review.